### PR TITLE
fix(server/player): use job/gang data as source of truth instead of playerdata

### PR DESCRIPTION
**Pull request by Manason**
The only thing the stored playerdata.job and .gang will be used for now is the name, grade, and last onduty value. Other fields of job/gang will be derived from the job/gang data. This fixes the potential for stale cached data on the player object. This doesn't change what data is written to the database.

### CODE REVIEW

**Review by solareon — APPROVED** [review]
Clean

File: server/player.lua [hunk 1]
```diff
@@ -390,32 +390,40 @@ function CheckPlayerData(source, playerData)
...
-    playerData.job.isboss = playerData.job.isboss or false
-    playerData.job.grade = playerData.job.grade or {}
-    playerData.job.grade.name = playerData.job.grade.name or 'Freelancer'
-    playerData.job.grade.level = playerData.job.grade.level or 0
+
+    local job = GetJob(playerData.job?.name) or GetJob('unemployed')
+    if not job then
+        error("unemployed job not found. Is it in your config and/or database?")
+    end
+    local jobGrade = GetJob(playerData.job?.name) and playerData.job.grade.level or 0
```
Comment (BerkieBb) [inline server/player.lua:398]: GetJob is redundant here and job grade should be grabbed from the `job` variable

File: server/player.lua [hunk 2]
```diff
@@ -390,32 +390,40 @@ function CheckPlayerData(source, playerData)
...
-    playerData.gang.label = playerData.gang.label or 'No Gang Affiliation'
-    playerData.gang.isboss = playerData.gang.isboss or false
-    playerData.gang.grade = playerData.gang.grade or {}
-    playerData.gang.grade.name = playerData.gang.grade.name or 'none'
-    playerData.gang.grade.level = playerData.gang.grade.level or 0
+    local gang = GetGang(playerData.gang?.name) or GetGang('none')
+    if not gang then
+        error("none gang not found. Is it in your config and/or database?")
+    end
+    local gangGrade = GetGang(playerData.gang?.name) and playerData.gang.grade.level or 0
```
Comment (BerkieBb) [inline server/player.lua:416]: Same here as job stuff

File: server/storage/main.lua [hunk 3]
```diff
@@ -24,7 +24,7 @@ local groups = require 'server.storage.groups'
 ---@field fetchIsUnique fun(type: UniqueIdType, value: string|number): boolean
 ---@field addPlayerToJob fun(citizenid: string, group: string, grade: integer)
 ---@field addPlayerToGang fun(citizenid: string, group: string, grade: integer)
----@field fetchPlayerGroups fun(citizenid: string): table<string, Job>, table<string, Gang>
+---@field fetchPlayerGroups fun(citizenid: string): table<string, integer>, table<string, integer> jobs, gangs
```
Comment (BerkieBb) [inline server/storage/main.lua:27]: Why this change and why the jobs, gangs at the end?

File: server/player.lua (hunk 1 above)
Comment (solareon) [reply server/player.lua:398]:
```suggestion
    local jobGrade = job.grade[playerData.job?.grade.level].level or 0
```
I believe this is what Berkie means. This should cover cases where a grade is deleted and the player has a grade that no longer exists

File: server/player.lua (hunk 2 above)
Comment (solareon) [reply server/player.lua:416]:
```suggestion
    local gangGrade = gang.grade[playerData.gang?.grade.level].level or 0
```

File: server/storage/main.lua (hunk 3 above)
Comment (Manason) [reply server/storage/main.lua:27]: Trying to communicate that the first table are jobs, and the second table are gangs. I guess the non lazy way to do it is with aliases.

This is a fix to make the documentation match the actual return value of the function

File: server/player.lua (hunk 1 above)
Comment (Manason) [reply server/player.lua:398]: the job variable is the actual job itself, not the player's job. Therefore it doesn't have player specific grade info, but rather all the grades of the job instead.

Comment (Manason) [reply server/player.lua:398]: > I believe this is what Berkie means. This should cover cases where a grade is deleted and the player has a grade that no longer exists

The reason why this won't work (aside from some off var names) is that job is either the saved job, or unemployed. If it's unemployed we want to disregard the saved player grade and go with 0.

Comment (solareon) [reply server/player.lua:398]: > > I believe this is what Berkie means. This should cover cases where a grade is deleted and the player has a grade that no longer exists
> 
> The reason why this won't work (aside from some off var names) is that job is either the saved job, or unemployed. If it's unemployed we want to disregard the saved player grade and go with 0.

My bad. I thought we were assembling the client side. disregard my comment

**Review by FjamZoo — APPROVED** [review]
